### ISSUE TITLE
Softmax Layer not failsafe

### ISSUE THREAD

**Issue by mkeller97** [Dobiasd/frugally-deep#90]
Hi Tobias,

First of all thank you for this great lib, it empowered me to go C++ with my keras models!
It works tremendously well for most of my setups, you are a genius!

However, I ran into the invalid divisor error in the softmax layer as others have before.
This happened just occasional and all the previous suggestions did not apply.
The Keras/Tensorflow/Python toolchain has no problems with the same data.

Debugging your lib I was able to locate the problem to  the input data going into the softmax layer.
The input for tensorflow is exactly the same but tensorflow can deal with it.

I found a suggestion for a failsafe implementation of softmax here:
[https://stackoverflow.com/questions/9906136/implementation-of-a-softmax-activation-function-for-neural-networks/9906882
](url)

let me illustrate the problem with some python script:
```
# softmax test
import numpy as np

# critical layer input is all high or all low like
# I take the numbers<0 from an actual example which runs in tf and fails in frugally deep
# all positive shows the same problem tehrefore I include the example
#inp = np.array([2139.21, 2407.21, 2205.09, 2178.34, 2693.79, 2210.82])
inp = np.array([-2139.21, -2407.21, -2205.09, -2178.34, -2693.79, -2210.82])

# exponentiation now gives all 0 or inf even with double precision!
outp = np.exp(inp)
s = np.sum(outp)
# now softmax fails as sum of output is inf or zero depending on the sign
prob = outp/s
print(prob)

# use the log-sum-exp trick to shift the input
# see https://stackoverflow.com/questions/9906136/implementation-of-a-softmax-activation-function-for-neural-networks/9906882
m = np.max(inp)
inp_shifted = inp-m
outp_shifted = np.exp(inp_shifted)
s_shifted = np.sum(outp_shifted)
# now softmax can be calculated in a failsafe manner
prob2 = np.exp(inp_shifted - np.log(s_shifted))
print(prob2)
#the result is pretty close to the result we get from tensorflow
```

I could even fix the problem for my use cases (input to softmax has format (1,1,6)) by modifying your code.

To give at least something back I thought to share it with you (forgive me if I didn't properly understand your code, you are on a different C++ level than I am)

The following is the modified snipet in softmax_layer.hpp

```
    tensor3 transform_input(const tensor3& input) const override
    {
        // Get unnormalized values of exponent function.
        const auto ex = [this](float_type x) -> float_type
        {
            return std::exp(x);
        };

        // --- previous std implementation by tobias herrmann
        //auto output = transform_tensor3(ex, input);

        // --- failsafe softmax by michael keller
        // shift input for failsafe softmax
        //float_type m = max_tensor3(input);
        float_type m = input.get(tensor3_max_pos(input));
        //auto inp_shift = input-m;
        auto inp_shift = subtract_tensor3(input,tensor3(input.shape(), m));
        auto outp_shifted = transform_tensor3(ex, inp_shift);

        // Softmax function is applied along channel dimension.
        for (size_t y = 0; y < input.shape().height_; ++y)
        {
            for (size_t x = 0; x < input.shape().width_; ++x)
            {
                // Get the sum of unnormalized values for one pixel.
                // We are not using Kahan summation, since the number
                // of object classes is usually quite small.
                //float_type sum = 0.0f;
                float_type sum_shifted = 0.0f;
                for (size_t z_class = 0; z_class < input.shape().depth_; ++z_class)
                {
                    // --- previous std implementation by tobias herrmann
                    //sum += output.get_yxz(y, x, z_class);
                    // --- failsafe softmax by michael keller
                    sum_shifted += outp_shifted.get_yxz(y, x, z_class);
                }
                /* michael keller: i do not see how the failsafe impl could give 0 or inf
                assertion(sum != 0 && !std::isinf(sum), "Invalid divisor in softmax layer. Try using\n"
                                                        "#define FDEEP_FLOAT_TYPE double\n"
                                                        "before\n"
                                                        "#include \"fdeep/fdeep.hpp\"\n"
                                                        "If you receive this error while using fdeep::load_model(filepath) try fdeep::load_model(filepath, false) instead to skip the verification step.");
                */
                // Divide the unnormalized values of each pixel by the stacks sum.
                for (size_t z_class = 0; z_class < input.shape().depth_; ++z_class)
                {
                    // --- previous std implementation by tobias herrmann
                    //output.set_yxz(y, x, z_class, output.get_yxz(y, x, z_class) / sum);
                    // --- failsafe softmax as output by michael keller
                    outp_shifted.set_yxz(y, x, z_class, std::exp(inp_shift.get_yxz(y, x, z_class)-std::log(sum_shifted)));
                }
            }
        }
        //return output;
        return outp_shifted;

    }

```

I guess it's not optimized and probably plain wrong but seems to work.
The python explanatory script tries to mimik the C++ and vice versa to make it simple to understand

If you want to take it or know of a better way to implement it, I'd be happy to see it in an upcoming version.

Feel free to ask if you would like more information, I'm happy to spend some more time debugging.

Regards,
Michael

**Comment by Dobiasd** [comment]
Hi Michael,

first, thanks a lot for these very nice words. :heart: 

And also thanks for this very good contribution. This softmax problem always bugged me a bit, however I did not yet find the right solution. But you did! :+1:

Also it seems like [TensorFlow really also does something shifting-related](https://github.com/tensorflow/tensorflow/blob/fc9c23a730ec9af278e5558ee4193801e36aa2ca/tensorflow/core/kernels/softmax_op_functor.h#L67). So that might be another point in favor of your implementation.

Would you like to commit your code changes as a pull request?

Tobias

**Comment by mkeller97** [comment]
Hi Tobias,

I'd be happy if you put it in, the way you think is appropriate.
Glad to have helped.

Currently I run testing if the results are equal in tensorflow and frugally-deep.
Up to now all looks good.

I will be able to spend some more time using your lib in the next few month so I'll keep you posted if I find more.

Keep up the good work!

Regards,
Michael

**Comment by Dobiasd** [comment]
OK cool. Just integrated your work into the code, successfully ran the tests, and [committed the changes](https://github.com/Dobiasd/frugally-deep/commit/b10040641bd199ea6c8d1d083b58884ff4fa4313).

Thanks again. :+1:

**Comment by Dobiasd** [comment]
Oh, btw, I also ran the performance measurements (from `README.md`) again, and there was no notable difference to the old version. :rocket:
